src/composition.rs

```diff
@@ -854,6 +879,37 @@ from(bucket: "myBucket")
         assert_eq!(expected, composition.to_string());
     }
 
+    /// A user can change the composition range and that range is retained
+    /// in the composition.
+    #[test]
+    fn test_composition_resolve_with_ast_range_change() {
+        let fluxscript = "".to_string();
+        let ast = flux::parser::parse_string("".into(), &fluxscript);
+
+        let mut composition = Composition::new(
+            ast,
+            "myBucket".into(),
+            Some("myMeasurement".into()),
+            vec![],
+            vec![],
+        );
+        let fluxscript = r#"from(bucket: "myBucket")
+    |> range(start: -24h)
+    |> filter(fn: (r) => r._measurement == "myMeasurement")
+"#;
+        let new_ast =
+            flux::parser::parse_string("".into(), &fluxscript);
+        assert!(composition.resolve_with_ast(new_ast).is_ok());
```

Comment:
My understanding is that every time there is an API call to LSP, a new AST is generated. Is it right?

Reply:
Every time we change something about composition, we generate a new AST, yes. In the case of `resolve_with_ast`, what we're trying to do is take an AST that was generated outside of composition and try and figure out where our composition statement is based on the state of composition and the new AST.